pep-9999.rst

```diff
@@ -0,0 +1,731 @@
+an inferred Literal type.
+
+For example, under what circumstances should literal expressions like ``"blue"``
+have an inferred type of ``Literal["blue"]`` vs ``str``?
+
+In short, type checkers are expected to be conservative and bias towards
+inferring standard types like ``str``. Type checkers should infer ``Literal[...]``
+only in contexts where a Literal type is explicitly requested.
+
+Variable assignment
+-------------------
+
+When assigning a literal expression to an unannotated variable, the
+inferred type of the variable is the original base type, not ``Literal[...]``.
+For example::
+
+    border_color = "blue"
+    reveal_type(border_color)   # Revealed type is 'str'
+
+This helps ensure we don't break the semantics of any existing code.
+
+If the user wants ``foo`` to have a literal type, they must 
+explicitly add an annotation::
+
+   border_color: Literal["blue"] = "blue"
+   reveal_types(border_color)   # Revealed type is 'Literal["blue"]'
+
+Or alternatively, use the ``Final`` qualifier::
+
+   border_color: Final = "blue"
+   reveal_types(border_color)   # Revealed type is 'Final[Literal["blue"]]'
+
+The ``Final`` qualifier will automatically infer a ``Literal`` type in
+an assignment if the LHS is a literal expression, or an expression of
+type ``Literal[...]``.
+
+**TODO:** Link to the PEP draft for the ``Final`` qualifier once it's ready.
+
+**Note 1:** A potential third way of declaring a Literal might be to 
+try using ``Literal`` as a qualifier::
+
+   foo: Literal = "hello"   # Illegal!
+
+Type checkers should *reject* lines like these. Unlike ``Final`` and ``ClassVar``,
+``Literal`` is a *type*, not a *qualifier*. Only qualifiers should infer their
+parameters.
+
+**Note 2:** Type checkers are only expected to use the context available
+to them within the current statement to infer the type of the variable.
+They may *optionally* use additional context to infer more precise types.
+For example::
+
+   def expects_blah(x: Literal[4]) -> None: ...
+
+   def test() -> None:
+       foo = 4
+       expects_blah(foo)
+
+In this program, it is theoretically possible for a type checker to deduce that
+``foo`` is only ever used as input to a function that expects ``Literal[4]``
+and so infer that ``foo`` must have type ``Literal[4]``.
+
+While type checkers *may* perform this kind of analysis, they are not obligated
+to do so. It is ok to infer that ``foo`` has type ``int``, since there is no
+context present in the assignment statement itself that would suggest that ``foo``
+should be a literal type.
+
+Type inference inside calls
+---------------------------
+
+When a literal is used inside of a function call, it will be inferred
+as either the original type or the Literal type based on context. For
+example, the following snippet is legal::
+
+   def expects_str(x: str) -> None: ...
+   def expects_literal(x: Literal["foo"]) -> None: ...
+
+   # Legal: "foo" is inferred to be of type 'str'
+   expects_str("foo")
```

Comment:
Couldn't inferring this to be a `Literal["foo"]` be fine, as that's a subclass of str?   I understand there are problems with containers due to variance, but that problem also exists for non-literals, and has also been unspecified so far.